### PR TITLE
perf: Materialize elements_chain

### PR DESCRIPTION
**Pull request by timgl**
## Problem

elements_chain is a huge column, and we use regex to parse it, so queries are slow. I looked at [the stats](https://metabase.prod-us.posthog.dev/question/799-action-autocapture-stats), and about half of actions have either text or href filters, so we can just extract those and query them directly, instead of through parsing elements_chain.

Of actions that only have selected, about half have matching on ID. Another 25% match on one of the 

## Changes

Add 4 materialized columns. I'll add the logic to actually query this later, that way we won't get "column doesn't exist" type issues.

👉 _Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review._

### CODE REVIEW

File: posthog/clickhouse/migrations/0064_materialize_elements_chain.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+ADD COLUMN IF NOT EXISTS elements_chain_mat_ids Array(String) MATERIALIZED arrayDistinct(extractAll(elements_chain, '(?::|\")id="(.*?)"')),
+ADD COLUMN IF NOT EXISTS elements_chain_mat_elements Array(Enum('a', 'button', 'form', 'input', 'select', 'textarea', 'label')) MATERIALIZED arrayDistinct(extractAll(elements_chain, '(?:^|;)(a|button|form|input|select|textarea|label)(?:\\.|$|:)'))
+"""
+
+
+def add_columns_to_required_tables(_):
+    with ch_pool.get_client() as client:
+        client.execute(ADD_COLUMNS_BASE_SQL.format(table="events", cluster=CLICKHOUSE_CLUSTER))
+
+        client.execute(ADD_COLUMNS_BASE_SQL.format(table="writable_events", cluster=CLICKHOUSE_CLUSTER))
```
Comment (fuziontech) [inline posthog/clickhouse/migrations/0064_materialize_elements_chain.py:27]: Even though you technically _can_ create `MATERIALIZED` columns on a distributed table like this now, this goes against how we do materialization. I'd take a look at `events` and `sharded_events` to see the difference, but the gist is you only create the materialized column on the `sharded` (actual MergeTree) table and then read through through that on the `events` distributed table. **You _do not_ want to put _any_ materialized columns or references to materialized columns on the writable table. This can break ingestion.**

You can see that on the `writable_events` table as it is on prod right now:
```sql
CREATE TABLE posthog.writable_events
(
    `uuid` UUID,
    `event` String,
    `properties` String,
    `timestamp` DateTime64(6, 'UTC'),
    `team_id` Int64,
    `distinct_id` String,
    `elements_hash` String,
    `created_at` DateTime64(6, 'UTC'),
    `_timestamp` DateTime,
    `_offset` UInt64,
    `elements_chain` String,
    `person_id` UUID,
    `person_properties` String,
    `group0_properties` String,
    `group1_properties` String,
    `group2_properties` String,
    `group3_properties` String,
    `group4_properties` String,
    `person_created_at` DateTime64(3),
    `group0_created_at` DateTime64(3),
    `group1_created_at` DateTime64(3),
    `group2_created_at` DateTime64(3),
    `group3_created_at` DateTime64(3),
    `group4_created_at` DateTime64(3),
    `inserted_at` Nullable(DateTime64(6, 'UTC')) DEFAULT NULL,
    `person_mode` Enum8('full' = 0, 'propertyless' = 1, 'force_upgrade' = 2)
)
ENGINE = Distributed('posthog', 'posthog', 'sharded_events', sipHash64(distinct_id))
```

`writable_events` represents the schema coming in from Kafka. All materialization occurs on the `MergeTree` table.

Basically you need to do two things here:
- Create the `MATERIALIZED` column on `sharded_events`
- Add a reference to it from `events`...something like 
```
`mat_ec_<column_name>` String COMMENT 'column_materializer_ec::<column_name>',
```

I used something similar here to how we do mat cols currently for properties, just added an `_ec` for elements chain. 

You don't need to change `writable_events`.

File: posthog/clickhouse/migrations/0064_materialize_elements_chain.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        client.execute(ADD_COLUMNS_BASE_SQL.format(table="sharded_events", cluster=CLICKHOUSE_CLUSTER))
+
+
+operations = [
+    run_sql_with_exceptions(f"DROP TABLE IF EXISTS events_json_mv ON CLUSTER '{CLICKHOUSE_CLUSTER}'"),
+    run_sql_with_exceptions(f"DROP TABLE IF EXISTS kafka_events_json ON CLUSTER '{CLICKHOUSE_CLUSTER}'"),
+    migrations.RunPython(add_columns_to_required_tables),
+    run_sql_with_exceptions(KAFKA_EVENTS_TABLE_JSON_SQL()),
+    run_sql_with_exceptions(EVENTS_TABLE_JSON_MV_SQL()),
```
Comment (fuziontech) [inline posthog/clickhouse/migrations/0064_materialize_elements_chain.py:36]: You don't need to do any operations with Kafka tables here.

You only need to add the column to:
- `sharded_events`
- `events`

No dropping of `kafka_` tables, no dropping of `_mv` tables.

File: posthog/models/event/sql.py [hunk 3]
```diff
@@ -22,6 +22,7 @@
     lambda: f"TRUNCATE TABLE IF EXISTS {EVENTS_DATA_TABLE()} ON CLUSTER '{settings.CLICKHOUSE_CLUSTER}'"
 )
 DROP_EVENTS_TABLE_SQL = lambda: f"DROP TABLE IF EXISTS {EVENTS_DATA_TABLE()} ON CLUSTER '{settings.CLICKHOUSE_CLUSTER}'"
+DROP_DISTRIBUTED_EVENTS_TABLE_SQL = f"DROP TABLE IF EXISTS events ON CLUSTER '{settings.CLICKHOUSE_CLUSTER}'"
```
Comment (fuziontech) [inline posthog/models/event/sql.py:25]: Drop this line - no need to drop `events` distributed table

**Review by fuziontech — CHANGES_REQUESTED** [review]
The logic looks solid here, but the way you are adding the materialized column is totally different from how we've done it in the past. I'd like to have this consistent since the way you are doing it here is actually more steps and will make the materialized schemas inconsistent.

File: posthog/models/event/sql.py (hunk 3 above)
Comment (timgl) [reply posthog/models/event/sql.py:25]: This is actually for the tests not production.

File: posthog/clickhouse/migrations/0064_materialize_elements_chain.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+ADD COLUMN IF NOT EXISTS mat_ec_ids Array(String) MATERIALIZED arrayDistinct(extractAll(elements_chain, '(?::|\")id="(.*?)"')),
+ADD COLUMN IF NOT EXISTS mat_ec_elements Array(Enum('a', 'button', 'form', 'input', 'select', 'textarea', 'label')) MATERIALIZED arrayDistinct(extractAll(elements_chain, '(?:^|;)(a|button|form|input|select|textarea|label)(?:\\.|$|:)'))
+"""
+
+ADD_COLUMNS_EVENTS = """
+ALTER TABLE {table} ON CLUSTER {cluster}
+ADD COLUMN IF NOT EXISTS mat_ec_href String COMMENT 'column_materializer::elements_chain::href',
+ADD COLUMN IF NOT EXISTS mat_ec_texts Array(String) COMMENT 'column_materializer::elements_chain::texts',
+ADD COLUMN IF NOT EXISTS mat_ec_ids Array(String) COMMENT 'column_materializer::elements_chain::ids',
+ADD COLUMN IF NOT EXISTS mat_ec_elements Array(Enum('a', 'button', 'form', 'input', 'select', 'textarea', 'label')) COMMENT 'column_materializer::elements_chain::elements'
```
Comment (fuziontech) [inline posthog/clickhouse/migrations/0064_materialize_elements_chain.py:20]: nice :sunglasses:

File: posthog/clickhouse/migrations/test/test_0064.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    for ec in elements_chains:
+        _create_event(
+            team=team,
+            distinct_id="1",
+            event="$autocapture",
+            elements=chain_to_elements(ec),
+        )
+
+
+class Test0064(ClickhouseDestroyTablesMixin):
```
Comment (fuziontech) [inline posthog/clickhouse/migrations/test/test_0064.py:25]: `ClickhouseDestroyTablesMixin` :yum:

File: posthog/models/event/sql.py (hunk 3 above)
Comment (fuziontech) [reply posthog/models/event/sql.py:25]: oops! yeah that makes sense :saluting_face:

**Review by fuziontech — APPROVED** [review]
Looks great!
Ship it :ship:
